preprocess/analysis/language_analysis_tests.py

```diff
@@ -0,0 +1,111 @@
+
+def entity_list(entities_as_tuple_list):
+  """Helper function for formatting expected entity analysis results.
+  
+  Properly formats a list of expected entity analysis results by mapping 
+  entity_dict across the input argument. 
+  
+  Args: 
+    entites_as_tuple_list: a 3-tuple list, where each tuple corresponds to 
+    an expected recognized entity, as name, type, salience.
+  
+  Returns:
+    a list of dictionaries, where each dictionary maps name, type, salience
+    to their correspoding string/string/float value.
+  """
+  
+  return list(map(lambda x: entity_dict(x), entities_as_tuple_list))
+
+def format_entity_list(entity_list):
+  """De-stringifies a list of json strings.
+  
+  Default return type for analysis is a json string. For testing it's 
+  convenient to just turn these back into dictionaries. 
+  
+  Args:
+    entity_list: a list of json strings, where each json string corresponds to
+    a recognized entity and its corresponding values (name, type, salience).
+    
+  Returns:
+    a list of dictionaries, where each dictionary maps name, type, salience
+    to their correspoding string/string/float value.
+  """
+  
+  return list(map(lambda x: json.loads(x), entity_list))
+
+class test_language_analysis(unittest.TestCase):
+  """Unit tests for language analysis
+  
+  Tests sentiment_analysis and entity_analysis with control values, and makes
+  sure that malformed inputs aren't accepted. 
+  """
+  
+  def test_string_one_sentiment(self):
+    expected_sentiment = sentiment_dict(0.1, 0.1)
+    actual_sentiment = json.loads(analyze_sentiment(STRING_ONE))
+    
+    self.assertEqual(expected_sentiment, actual_sentiment)
+    
+  def test_string_one_entities(self):
+    expected_entity_list = [("Pro Wrestler", "ORGANIZATION", 0.62), 
+                           ("Congress", "ORGANIZATION", 0.19),
+                           ("Rebuild Nevada", "OTHER", 0.13), 
+                           ("Big Dan Rodimer", "PERSON", 0.06)]
+    formatted_expected_entity_list = entity_list(expected_entity_list)
+    
+    actual_list = format_entity_list(analyze_entities(STRING_ONE))
+
+    self.assertEqual(formatted_expected_entity_list, actual_list)
+    
+  def test_string_two_sentiment(self):
+    expected_sentiment = sentiment_dict(-0.4, 0.4)
+    actual_sentiment = json.loads(analyze_sentiment(STRING_TWO))
+    
+    self.assertEqual(expected_sentiment, actual_sentiment)
+    
+  def test_string_two_entities(self):
+    expected_entity_list = [("John Bel Edwards", "PERSON", 0.41),
+                            ("LA", "LOCATION", 0.29),
+                            ("No Sanctuary Cities", "LOCATION", 0.22),
+                            ("It's Up to You", "WORK_OF_ART", 0.09)]
+    formatted_expected_entity_list = entity_list(expected_entity_list)
+    
+    actual_list = format_entity_list(analyze_entities(STRING_TWO))
+
+    self.assertEqual(formatted_expected_entity_list, actual_list)
+    
```

Comment:
You might want to add some tests where the entities aren't formatted correctly and make sure those cases are handled appropriately

Reply:
+1.  I'll pay you $1,000,000 to break this code.  How do you do it?

Think about that, then write those test cases.

Reply:
See discussion here https://github.com/googleinterns/step9-2020/pull/2#discussion_r448584805

Tl;dr:
- the next level of useful testing is probably out of scope and would require some kind of language detection. 
- as long as the input is a valid python string, it's going to be a valid input for the API call.
- encoding  type is a non-issue.